### PR TITLE
swev-id: django__django-13089 guard DatabaseCache cull pivot

### LINKED ISSUE #212: DatabaseCache._cull fails with 'NoneType' object is not subscriptable when pivot SELECT returns no rows
## User report

Cache backend: database. Sporadic errors observed using Django 2.2.11 (same code path present on master).

Error: `'NoneType' object is not subscriptable`

Stack trace:

```
/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py:143 → _get_response
/usr/local/lib/python3.7/site-packages/django/template/response.py:108 → render
/usr/local/lib/python3.7/site-packages/django/utils/decorators.py:156 → callback
/usr/local/lib/python3.7/site-packages/django/middleware/cache.py:103 → process_response
/usr/local/lib/python3.7/site-packages/django/utils/cache.py:374 → learn_cache_key
/usr/local/lib/python3.7/site-packages/django/core/cache/backends/db.py:104 → set
/usr/local/lib/python3.7/site-packages/django/core/cache/backends/db.py:136 → _base_set
/usr/local/lib/python3.7/site-packages/django/core/cache/backends/db.py:277 → _cull
```

Source reference:
https://github.com/django/django/blob/master/django/core/cache/backends/db.py#L270

```
cursor.execute(
    connection.ops.cache_key_culling_sql() % table,
    [cull_num]
)
cursor.execute("DELETE FROM %s WHERE cache_key < %%s" % table, [cursor.fetchone()[0]])
```

Observation: After executing `connection.ops.cache_key_culling_sql()` the cursor may return no rows; `cursor.fetchone()` is `None`, causing the subsequent index `[0]` to raise.

## Reproduction steps

1. Configure DatabaseCache with a small MAX_ENTRIES and CULL_FREQUENCY = 1:
   ```python
   CACHES = {
       'default': {
           'BACKEND': 'django.core.cache.backends.db.DatabaseCache',
           'LOCATION': 'test cache table',
           'OPTIONS': {
               'MAX_ENTRIES': 2,
               'CULL_FREQUENCY': 1,
           },
       },
   }
   ```
2. Create the cache table (`manage.py createcachetable`).
3. Insert more than MAX_ENTRIES keys to trigger culling:
   ```python
   for i in range(1, 6):
       cache.set(f'k{i}', i, timeout=1000)
   ```
4. On overflow, `_cull()` computes `cull_num = num // CULL_FREQUENCY` which equals `num`; the pivot query uses `OFFSET num` and returns no rows. `cursor.fetchone()[0]` then raises `'NoneType' object is not subscriptable'.

## Specification (proposed fix and tests)

- Guard culling pivot selection against out-of-range OFFSET and None rows.
- When `cull_num >= num` (e.g., CULL_FREQUENCY = 1), clear the table (full cull) and return.
- Otherwise, select pivot and delete rows below it; if pivot row is None (race), return safely.
- Add tests:
  - CULL_FREQUENCY = 1 scenario: insert > MAX_ENTRIES; ensure no exception and final count == 1 (last key).
  - Defensive None pivot: if pivot query returns no row, no exception; the new key is present.
  - Ensure default culling tests still pass (CULL_FREQUENCY = 3).

This issue tracks implementation on branch django__django-13089.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- guard DatabaseCache._cull when cull frequency would remove all rows
- add regression coverage for MAX_ENTRIES=2 and CULL_FREQUENCY=1
- satisfy lint in cache test teardown

## Reproduction
1. Configure DatabaseCache with MAX_ENTRIES=2 and CULL_FREQUENCY=1.
2. Insert four distinct keys to force a cull.
3. Observe previous `'NoneType' object is not subscriptable` error; new logic clears the table instead.

## Testing
- PYTHONPATH=/workspace/django ./runtests.py cache.tests --parallel=1
- flake8 django/core/cache/backends/db.py tests/cache/tests.py

Fixes #212

### CODE REVIEW

**Comment by casey-brooks** [comment]
Test & Lint Summary
- PYTHONPATH=/workspace/django ./runtests.py cache.tests --parallel=1 (478 passed, 0 failed, 115 skipped)
- flake8 django/core/cache/backends/db.py tests/cache/tests.py (no issues)

**Comment by rowan-stein** [comment]
Requesting review:

This PR fixes a sporadic crash in DatabaseCache._cull when the pivot query returns no rows (e.g., CULL_FREQUENCY=1 causing OFFSET == row count). Changes:
- Clear table when cull_num >= num and handle None pivot rows defensively.
- Add regression tests for MAX_ENTRIES=2 / CULL_FREQUENCY=1; existing cull tests remain green.

Focus areas:
- Correctness of cull behavior across frequencies.
- Defensive handling of race conditions between COUNT and pivot SELECT.
- Test coverage and expectations.

swev-id: django__django-13089

File: django/core/cache/backends/db.py [hunk 1]
```diff
@@ -264,12 +264,20 @@ def _cull(self, db, cursor, now):
...
                 cursor.execute(
                     connection.ops.cache_key_culling_sql() % table,
-                    [cull_num])
-                cursor.execute("DELETE FROM %s "
-                               "WHERE cache_key < %%s" % table,
-                               [cursor.fetchone()[0]])
+                    [cull_num]
+                )
+                pivot_row = cursor.fetchone()
+                if pivot_row is None:
```
Comment (noa-lucent) [inline django/core/cache/backends/db.py:275]: [major] Guarding the new `pivot_row is None` branch removes the crash, but we need regression coverage to keep it that way. Could you add a test (e.g., patching the cursor or `_cull()` to return `None` once) so we verify this path? The issue report specifically calls out this scenario, so we should lock it down with a test.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for tackling the DatabaseCache culling edge cases. I left one blocker: we need regression coverage for the new  guard so the fix doesn’t regress.

**Comment by casey-brooks** [comment]
Follow-up on Noa's review: added regression coverage for the `_cull()` guard when the culling query returns no pivot row. Ran `PYTHONPATH=$PWD:$HOME/.nix-profile/lib/python3.12/site-packages DJANGO_SETTINGS_MODULE=tests.test_sqlite TZDIR=/usr/share/zoneinfo python tests/runtests.py cache --parallel=1` plus targeted utils modules (`... utils_tests.test_datetime_safe`, `... utils_tests.test_hashable`). The full `utils_tests` label still fails on this runner with the existing cached_property/module_loading/timezone/htmlparser issues, so I captured the traces above and will revisit once the environment alignment work is done.
